promql/parse.go

```diff
@@ -500,7 +500,8 @@ func (p *parser) expr() Expr {
 }
 
 func (p *parser) balance(lhs Expr, op itemType, rhs Expr, vecMatching *VectorMatching, returnBool bool) *BinaryExpr {
-	if lhsBE, ok := lhs.(*BinaryExpr); ok && lhsBE.Op.precedence() < op.precedence() {
+	if lhsBE, ok := lhs.(*BinaryExpr); (ok && lhsBE.Op.precedence() < op.precedence()) ||
+		(ok && lhsBE.Op.precedence() == op.precedence() && lhsBE.Op == itemPOW && op == itemPOW) {
```

Comment:
Would indeed be good to encapsulate associativity logic here for future clarity. We could define another method similar to `precedence()`; `isRightAssociative()` seems simple enough.

To make the code more self-explanatory we could structure this condition something like:

```go
lhsBE, ok := lhs.(*BinaryExpr)
precd := lhsBe.Op.precedence() - op.precedence()
// Relies on the invariant that no operators with the same precedence level can have different associativity.
if ok && (precd < 0 || precd == 0 && op.isRightAssociative()) {
    // ...
}
```

Does that make sense?